### DISCUSSION TITLE
How does one edit the passes XLA uses in compilation?

### DISCUSSION THREAD

**Discussion by IcebladeLabs** [openxla/community#122] [xla]
I was able to find a file which added passes to pipeline, but I was curious as to the correct way to add/ remove passes used in compilation. Thank you for your time!

**Comment by abhigunj** [comment]
Hi @IcebladeLabs Thanks for your question! 
It depends on what you are trying to do and type of the pass. If you know what you are doing (aware of pipeline, pass configs, interdependencies) then you can directly modify the pipeline by add/removing pipeline.AddPass(..). Otherwise take a look at following options: 

1.  If you want to play with the existing pipeline,  add / remove passes then use following options to enable/disable the passes.
```
hlo-opt --platform=<> --stage=hlo --xla_enable_hlo_passes_only=<pass_names> input.hlo
hlo-opt --platform=<> --stage=hlo --xla_disable_hlo_passes =<pass_names> input.hlo
```
Limitation: This option may not work if a pipeline become invalid because inter-dependent passes are disabled.

2.  If you want to decouple from existing pipeline and experiment with new or existing single or set of passes, we are adding new enhancements to `hlo-opt` at https://github.com/openxla/xla/pull/18935 (ETA: End of this Week). It will allow you to run a single or set of passes without worrying about existing pipelines.
```
hlo-opt -- --passes=<pass_name1,pass_name2>
```
Limitation: currently only hardware independent passes, which does not accept any special configs are supported.

Let us know if this helps. Feel free to share more details about pass and pipeline you are working with.
